rule/add_constant.go

```diff
@@ -39,14 +39,14 @@ type AddConstantRule struct {
 	strLitLimit     int
 
 	configureOnce sync.Once
+	configureErr  error
 }
 
 // Apply applies the rule to given file.
 func (r *AddConstantRule) Apply(file *lint.File, arguments lint.Arguments) []lint.Failure {
-	var configureErr error
-	r.configureOnce.Do(func() { configureErr = r.configure(arguments) })
-	if configureErr != nil {
-		return newInternalFailureError(configureErr)
+	r.configureOnce.Do(func() { r.configureErr = r.configure(arguments) })
+	if r.configureErr != nil {
```

Comment:
No risk of a data race on `r.configureErr` between:
1. the first thread doing a write at `r.configureErr = r.configure(arguments)` and
2. a second thread doing a read at `r.configureErr != nil` 
?

Reply:
You're right. We can add a test to check these thoughts.

Reply:
I'm not sure that adding tests is enough, the approach itself is inherently data-race-prone: adding the field `configureErr` re-creates the problem (data-races) we solved with a mutex (later replaced by `sync.Once`)

IMO, it's `sync.Once` that makes things (error management) complicated. An approach combining a mutex and `configureErr` might provide a flawless solution (at the price of verbosity):

```go
func (r *TheRule) configure () error {
r.Lock()
defer r.Unlock

if r.alreadyConfigured {
  return r.configureErr
}
r.alreadyConfigured = true

// ...
r.configureErr = ...
```

Reply:
Maybe even better is to split rule configuration and rule applying. In that case, we don't need neither mutex nor sync.Once.

Reply:
Is an alternative, we can define a `configurable` interface and we call `r.Configure` on each rule implementing the interface.

Reply:
This is a very good idea. I can tell because I suggested it too, a few days ago :rofl: 

https://github.com/mgechev/revive/pull/1178#discussion_r1874804066

I think I will move into this direction.

Reply:
OK, I implemented it.